State.cpp

```diff
@@ -0,0 +1,248 @@
+#include "BreathData.h"
+#include "State.h"
+#include "Constants.h"
+#include "Display.h"
+#include "Flow.h"
+#include "Pressure.h"
+#include "Valve.h"
+#include "ProportionalValve.h"
+#include "AlarmManager.h"
+#include "State.h"
+
+// Initialize singleton instances
+OffState* OffState::instance = 0;
+InsStateVC* InsStateVC::instance = 0;
+InsHoldStateVC* InsHoldStateVC::instance = 0;
+ExpStateVC* ExpStateVC::instance = 0;
+PeepStateVC* PeepStateVC::instance = 0;
+ExpHoldStateVC* ExpHoldStateVC::instance = 0;
+
+BreathData* State::breath;
+
+/*********************************************************/
+/* OffState                                              */
+/*********************************************************/
+
+// performs actions necessary upon entering a state
+static State* OffState::begin(BreathData b) {
+    if (instance==0) instance = new OffState();  // one-time setup
+    breath = &b;                                 // store reference to breath data
+
+    // valve settings
+    inspValve.endBreath(); // close the inspiratory valve
```

Comment:
This is confusing. Why are there actions in something that supposedly just returns the address of a singleton?  Don't mix up accessors with action functions.

Reply:
There aren't really any pure accessors here.  The idea is for the program to keep just one pointer to the current state.  The only way to get a pointer to another state is to enter it by calling `enter`.

I suppose one could take this idea further and make the state instances completely unavailable to the main program, which would then only call static methods in State.  Not sure that's an improvement, though.